database/tests/desktop/core/sync_point_spec_test.cc

```diff
@@ -0,0 +1,883 @@
+#include "database/src/desktop/persistence/persistence_manager_interface.h"
+#include "database/src/desktop/persistence/persistence_storage_engine.h"
+#include "database/src/desktop/query_desktop.h"
+#include "database/src/desktop/util_desktop.h"
+#include "database/src/include/firebase/database/common.h"
+#include "flatbuffers/util.h"
+#include "gmock/gmock.h"
+#include "gtest/gtest.h"
+
+using ::testing::Eq;
+using ::testing::Not;
+using ::testing::Pointwise;
+using ::testing::Test;
+using ::testing::UnorderedPointwise;
+
+namespace firebase {
+
+using util::FlexbufferToVariant;
+
+namespace database {
+namespace internal {
+
+// Compare all fields of Events except the event registration pointer.
+MATCHER(EventEq, "EventEq") {
+  const Event& event_a = std::get<0>(arg);
+  const Event& event_b = std::get<1>(arg);
+  if (event_a.snapshot.has_value() != event_b.snapshot.has_value()) {
+    return false;
+  } else if (event_a.snapshot.has_value()) {
+    const DataSnapshotInternal& snapshot_a = *event_a.snapshot;
+    const DataSnapshotInternal& snapshot_b = *event_b.snapshot;
+    if (snapshot_a.GetKeyString() != snapshot_b.GetKeyString() ||
+        snapshot_a.GetValue() != snapshot_b.GetValue() ||
+        snapshot_a.GetPriority() != snapshot_b.GetPriority()) {
+      return false;
+    }
+  }
+  return event_a.prev_name == event_b.prev_name && event_a.path == event_b.path;
+}
+
+namespace {
+
+class FakeListenProvider : public ListenProvider {
+ public:
+  FakeListenProvider(LoggerBase* logger) : logger_(logger), listens_() {}
+
+  ~FakeListenProvider() override {}
+
+  void StartListening(const QuerySpec& query_spec, const Tag& tag,
+                      const View* view) override {
+    const Path& path = query_spec.path;
+    logger_->LogDebug(
+        "Listening at %s for Tag %s", path.c_str(),
+        tag.has_value() ? std::to_string(*tag).c_str() : "<None>");
+    assert(listens_.count(query_spec) == 0);
+    listens_.insert(query_spec);
+  }
+
+  void StopListening(const QuerySpec& query_spec, const Tag& tag) override {
+    const Path& path = query_spec.path;
+    logger_->LogDebug(
+        "Stop listening at %s for Tag %s", path.c_str(),
+        tag.has_value() ? std::to_string(*tag).c_str() : "<None>");
+    assert(listens_.count(query_spec) > 0);
+    listens_.erase(query_spec);
+  }
+
+ private:
+  LoggerBase* logger_;
+  std::set<QuerySpec> listens_;
+};
+
+class SyncTreeTest : public Test {
+ public:
+  SyncTreeTest() {
+    const char* kTestDataFile = "sync_point_spec.bin";
+    flatbuffers::LoadFile(kTestDataFile, true, &buffer_);
+    test_suite_ = test_data::GetTestSuite(buffer_.data());
+  }
+
+  void SetUp() override {
+    sync_tree_ = new SyncTree(MakeUnique<WriteTree>(),
+                              MakeUnique<NoopPersistenceManager>(),
+                              MakeUnique<FakeListenProvider>(&logger_));
+  }
+
+  void TearDown() override { delete sync_tree_; }
+  void runOne(const char* name);
```

Comment:
nit: why aren't these named RunOne and RunTest?

Reply:
No good reason, it was an oversight. Fixed.